video/vdu_buffered.h

```diff
@@ -1060,4 +1072,203 @@ void VDUStreamProcessor::bufferCopyAndConsolidate(uint16_t bufferId, std::vector
 	debug_log("bufferCopyAndConsolidate: copied %d bytes into buffer %d\n\r", length, bufferId);
 }
 
+// Write compressed output to a temporary buffer
+//
+static void local_write_compressed_byte(void* p_cd, uint8_t comp_byte) {
+	CompressionData* cd = (CompressionData*) p_cd;
+    uint8_t** pp_temp = (uint8_t**) cd->context;
+	uint8_t* p_temp = *pp_temp;
+	p_temp[cd->output_count++] = comp_byte;
+	if (!(cd->output_count & (COMPRESSION_OUTPUT_CHUNK_SIZE-1))) {
+		// extend the temporary buffer to make room for more output
+		auto new_size = cd->output_count + COMPRESSION_OUTPUT_CHUNK_SIZE;
+		uint8_t* p_temp2 = (uint8_t*) ps_malloc(new_size);
+		if (p_temp2) {
+			memcpy(p_temp2, p_temp, cd->output_count);
+			*pp_temp = p_temp2; // use new buffer pointer now
+		} else {
+			debug_log("bufferCompress: cannot allocate temporary buffer of %d bytes\n\r", new_size);
+			*pp_temp = NULL; // indicate failure
+		}
+		heap_caps_free(p_temp);
+	}
+}
+
+// Write decompressed output to a temporary buffer
+//
+static void local_write_decompressed_byte(void* p_dd, uint8_t orig_data) {
+	DecompressionData* dd = (DecompressionData*) p_dd;
+    uint8_t** pp_temp = (uint8_t**) dd->context;
+	uint8_t* p_temp = *pp_temp;
+	p_temp[dd->output_count++] = orig_data;
+	if (!(dd->output_count & (COMPRESSION_OUTPUT_CHUNK_SIZE-1))) {
+		// extend the temporary buffer to make room for more output
+		auto new_size = dd->output_count + COMPRESSION_OUTPUT_CHUNK_SIZE;
+		uint8_t* p_temp2 = (uint8_t*) ps_malloc(new_size);
+		if (p_temp2) {
+			memcpy(p_temp2, p_temp, dd->output_count);
+			*pp_temp = p_temp2; // use new buffer pointer now
+		} else {
+			debug_log("bufferDecompress: cannot allocate temporary buffer of %d bytes\n\r", new_size);
+			*pp_temp = NULL; // indicate failure
+		}
+		heap_caps_free(p_temp);
+	}
+}
+
+// VDU 23, 0, &A0, bufferId; &1C, sourceBufferId; sourceBufferId; ...; 65535; : Compress blocks from buffers
```

Comment:
comment needs updating to reflect the fact that this takes a single source buffer 😁

Reply:
Oops. Nice catch.